wandb/sdk/launch/environment/aws_environment.py

```diff
@@ -0,0 +1,238 @@
+                destination is not a valid s3 URI, or the upload fails.
+        """
+        _logger.debug(f"Uploading {source} to {destination}")
+        if not os.path.isfile(source):
+            raise LaunchError(f"Source {source} does not exist.")
+        match = s3_uri_re.match(destination)
+        if not match:
+            raise LaunchError(f"Destination {destination} is not a valid s3 URI.")
+        bucket = match.group(1)
+        key = match.group(2).lstrip("/")
+        if not key:
+            key = ""
+        session = self.get_session()
+        try:
+            client = session.client("s3")
+            client.upload_file(source, bucket, key)
+        except botocore.exceptions.ClientError as e:
+            raise LaunchError(
+                f"botocore error attempting to copy {source} to {destination}. {e}"
+            )
+
+    def upload_dir(self, source: str, destination: str) -> None:
+        """Upload a directory to s3 from local storage.
+
+        The upload will place the contents of the source directory in the destination
+        with the same directory structure. So if the source is "foo/bar" and the
+        destination is "s3://bucket/key", the contents of "foo/bar" will be uploaded
+        to "s3://bucket/key/bar".
+
+        Args:
+            source (str): The path to the file or directory.
+            destination (str): The URI of the storage.
+            recursive (bool, optional): If True, copy the directory recursively. Defaults to False.
+
+        Raises:
+            LaunchError: If the copy fails, the source path does not exist, or the
+                destination is not a valid s3 URI.
+        """
+        _logger.debug(f"Uploading {source} to {destination}")
+        if not os.path.isdir(source):
+            raise LaunchError(f"Source {source} does not exist.")
+        match = s3_uri_re.match(destination)
+        if not match:
+            raise LaunchError(f"Destination {destination} is not a valid s3 URI.")
+        bucket = match.group(1)
+        key = match.group(2).lstrip("/")
+        if not key:
+            key = ""
+        session = self.get_session()
+        try:
+            client = session.client("s3")
+            for path, _, files in os.walk(source):
+                for file in files:
+                    abs_path = os.path.join(path, file)
+                    key_path = (
+                        abs_path.replace(source, "").replace("\\", "/").lstrip("/")
+                    )
+                    client.upload_file(
+                        abs_path,
+                        bucket,
+                        key_path,
+                    )
+        except botocore.exceptions.ClientError as e:
+            raise LaunchError(
+                f"botocore error attempting to copy {source} to {destination}. {e}"
+            )
+        except Exception as e:
+            raise LaunchError(
```

Comment:
from e

Reply:
Do we know when this error can happen?

Reply:
I just put this as a blanket in case something not expected happened, like for some reason a file in the walk no longer existed when we tried to upload.

Reply:
Great!